src/ffi.rs

```diff
@@ -4,686 +4,621 @@
+        }
+
+        // optimizes the inputs given the total amount
+        let (unspent, inputs) = utils::knapsack(full_inputs, total_output)?;
+        let inputs: Vec<_> = inputs
+            .into_iter()
+            .map(|(note, opening, value, idx)| tx::PreInput {
+                note,
+                opening,
+                value,
+                ssk: &keys_ssk[idx],
+            })
+            .collect();
+        let total_input: u64 = inputs.iter().map(|i| i.value).sum();
+        let total_refund = total_input.saturating_sub(total_output);
+
+        let mut outputs: Vec<tx::OutputValue> = Vec::with_capacity(2);
+        if let Some(o) = output {
+            outputs.push(o);
+        }
+        if total_refund > 0 {
+            outputs.push(tx::OutputValue {
+                r#type: NoteType::Obfuscated,
+                value: total_refund,
+                receiver: refund,
+                ref_id: 0,
+            });
+        }
+
+        let rng = &mut utils::rng(&rng_seed);
+        let tx = tx::UnprovenTransaction::new(
+            rng, inputs, outputs, &refund, gas_limit, gas_price, crossover,
+            call,
+        )?;
+
+        let unspent = rkyv::to_bytes::<_, MAX_LEN>(&unspent).ok()?.into_vec();
+        let tx = rkyv::to_bytes::<_, MAX_LEN>(&tx).ok()?.into_vec();
+
+        Some((unspent, tx))
+    }
+
+    let (unspent, tx) = match inner(args) {
+        Some(t) => t,
+        None => return ExecuteResponse::fail(),
+    };
+
+    let unspent_ptr = unspent.as_ptr() as u64;
+    let unspent_len = unspent.len() as u64;
+    let tx_ptr = tx.as_ptr() as u64;
+    let tx_len = tx.len() as u64;
+
+    mem::forget(unspent);
+    mem::forget(tx);
+
+    ExecuteResponse::success(unspent_ptr, unspent_len, tx_ptr, tx_len)
 }
 
-/// Unstake the value previously staked using the [`stake`] function.
+/// Merges many lists of serialized notes into a unique, sanitized set.
+///
+/// The arguments are expected to be rkyv serialized [MergeNotesArgs] with a
+/// pointer defined via [malloc]. It will consume the `args` allocated region
+/// and drop it.
 #[no_mangle]
-pub unsafe extern "C" fn unstake(
-    sender_index: *const u64,
-    staker_index: *const u64,
-    refund: *const [u8; PublicSpendKey::SIZE],
-    gas_limit: *const u64,
-    gas_price: *const u64,
-) -> u8 {
-    let refund = unwrap_or_bail!(PublicSpendKey::from_bytes(&*refund));
-
-    unwrap_or_bail!(WALLET.unstake(
-        &mut FfiRng,
-        *sender_index,
-        *staker_index,
-        &refund,
-        *gas_price,
-        *gas_limit
-    ));
-
-    0
+pub fn merge_notes(args: i32, len: i32) -> i32 {
+    let args = args as *mut u8;
+    let len = len as usize;
+    let args = unsafe { Vec::from_raw_parts(args, len, len) };
+
+    let MergeNotesArgs { notes } = match rkyv::from_bytes(&args) {
+        Ok(a) => a,
+        Err(_) => return MergeNotesResponse::fail(),
+    };
+
+    let len = 3 * notes.len() / mem::size_of::<ArchivedNote>() / 2;
```

Comment:
Just checking: We don't need any overflow checks here right?

Reply:
It would overflow only if we have a massive vec of notes; but OTOH this was indeed more complicated than it should be. I reworked this into a simpler, more readable implementation